core/traffic_class.h

```diff
@@ -376,16 +376,36 @@ class RateLimitTrafficClass final : public TrafficClass {
   void FinishAndAccountTowardsRoot(Scheduler *sched, TrafficClass *child,
                                    resource_arr_t usage, uint64_t tsc) override;
 
-  inline resource_t resource() const { return resource_; }
+  resource_t resource() const { return resource_; }
 
-  inline uint64_t limit() const { return limit_; }
+  // Return the configured limit, in work units
+  uint64_t limit() const { return limit_; }
 
-  inline uint64_t max_burst() const { return max_burst_; }
+  // Return the configured max burst, in work units
+  uint64_t max_burst() const { return max_burst_; }
+
+  void set_resource(resource_t res) { resource_ = res; }
+
+  // Set the limit to `limit`, which is in units of the resource type
+  void set_limit(uint64_t limit) { limit_ = to_work_units(limit); }
+
+  // Set the max burst to `burst`, which is in units of the resource type
+  void set_max_burst(uint64_t burst) { max_burst_ = to_work_units(burst); }
 
   TrafficClass *child() const { return child_; }
 
   void Traverse(TravereseTcFn f, void *arg) const override;
 
+  // Convert resource units to work units
+  static uint64_t to_work_units(uint64_t x) {
+    return (x << (USAGE_AMPLIFIER_POW - 4)) / (tsc_hz >> 4);
+  }
+
+  // Convert work units to resource units
+  static uint64_t to_resource_units(uint64_t x) {
+    return (x * (tsc_hz >> 4)) >> (USAGE_AMPLIFIER_POW - 4);
+  }
+
```

Comment:
Yeah, Melvin and I discussed it.  It's truncation, and the +1 hack is not a good option.  We agreed the best option is to store the original value.